File: swf-animate/manager/FxAtlasManager.cpp
//
//  FxAtlasManager.cpp
//
//  Created by joli on 2018/8/14.
//  Copyright ? 2018Äê uzone. All rights reserved.
//

#include "flashx/manager/FxAtlasManager.h"

NS_FLASHX_BEGIN

#define TEX_FORMAT ".png"
#define insSpFrameCache() SpriteFrameCache::getInstance()
#define insTextureCache() Director::getInstance()->getTextureCache()

class FxAtlasItem : public Ref {
public:
    inline void pushFrame(SpriteFrame* sprFrame) { frames.pushBack(sprFrame); }
private:
    cocos2d::Vector<SpriteFrame*> frames;
};

static fxstr getImagePath(const fxstr& plistPath)
{
    fxstr path = plistPath;
    const size_t& pos = path.find_last_of(".");
    return path.erase(pos).append(TEX_FORMAT);
}

FxAtlasManager* FxAtlasManager::getInstance() {
    static FxAtlasManager instance;
    return &instance;
}

FxAtlasManager::FxAtlasManager() {}

FxAtlasManager::~FxAtlasManager() {}

bool FxAtlasManager::retainAtlas(const fxstr& plistPath, Texture2D* texture /* nullptr */)
{
    auto atlasIter = atlasMap.find(plistPath);
    if (atlasMap.end() != atlasIter) {
        //FXLOG("%s:%s", __func__, plistPath.c_str());
        atlasIter->second->retain();
        return true;
    }
    if (!texture) {
        const fxstr& imagePath = getImagePath(plistPath);
        texture = insTextureCache()->addImage(imagePath);
    }
    if (texture) {
        FxAtlasItem *atlasItem = putAtlas(plistPath, texture);
        if (atlasItem) {
            //FXLOG("%s:%s", __func__, plistPath.c_str());
            atlasItem->retain();
            return true;
        }
    }
    FXLOG("%s load texture failed:%s", __func__, plistPath.c_str());
    return false;
}

void FxAtlasManager::releaseAtlas(const fxstr& plistPath)
{
    auto atlasIter = atlasMap.find(plistPath);
    if (atlasMap.end() == atlasIter) {
        return;
    }
    FxAtlasItem *atlasItem = atlasIter->second;
    if (atlasItem) {
        //FXLOG("%s:%s", __func__, plistPath.c_str());
        atlasItem->release();
    }
}

void FxAtlasManager::removeUnsuedAtlas()
{
    std::vector<fxstr> removes;
    forit(atlasMap, altasItr) {
        if (altasItr->second->getReferenceCount() == 1) {
            removes.push_back(altasItr->first);
        }
    }
    
    for(const auto &plistPath : removes) {
        atlasMap.erase(plistPath);
        insSpFrameCache()->removeSpriteFramesFromFile(plistPath);// remove spframes
        insTextureCache()->removeTextureForKey(getImagePath(plistPath));// remove texture
    }
    
//    atlasMap.erase(removes);
//    insSpFrameCache()->removeUnusedSpriteFrames();
//    insTextureCache()->removeUnusedTextures();
    
    // FXLOG("%s:%lu", __func__, removes.size());
}

FxAtlasItem* FxAtlasManager::putAtlas(const fxstr& plistPath, Texture2D* texture)
{
    SpriteFrameCache *sfc = insSpFrameCache();
    ValueMap plistDict = sfc->getPlistDataLocal(plistPath);
    if (plistDict.empty()) {
        return nullptr;
    }
    
    // FXLOG("%s:%s", __func__, plistPath.c_str());
    sfc->addSpriteFramesWithFile(plistPath, texture);
    FxAtlasItem *atlasItem = new FxAtlasItem();
    if (atlasItem) {
        const ValueMap& frameDict = plistDict["frames"].asValueMap();
        forit(frameDict, iter) {
            atlasItem->pushFrame(sfc->getSpriteFrameByName(iter->first));
        }
        atlasMap.insert(plistPath, atlasItem);
        atlasItem->release();
        return atlasItem;
    }
    
    FX_SAFE_DELETE(atlasItem);
    return nullptr;
}

void FxAtlasManager::loadAtlasTexture(const fxstr& plistPath, const loadTexHandler& callback)
{
    const fxstr& imagePath = getImagePath(plistPath);
    // FXLOG("%s load texture:%s", __func__, imagePath.c_str());
    insTextureCache()->addImageAsync(imagePath, callback);
}

Texture2D* FxAtlasManager::loadImageByFormat(const fxstr& imagePath,
                                             const loadTexHandler& callback /* nullptr */,
                                             Texture2D::PixelFormat format /* Texture2D::PixelFormat::RGBA4444 */)
{
    const Texture2D::PixelFormat& def = Texture2D::getDefaultAlphaPixelFormat();
    Texture2D::setDefaultAlphaPixelFormat(format);
    // async
    if (callback) {
        insTextureCache()->addImageAsync(imagePath, [callback, def](Texture2D* texture) {
            Texture2D::setDefaultAlphaPixelFormat(def);
            callback(texture);
        });
        return nullptr;
    }
    // synchronized
    Texture2D* texture = insTextureCache()->addImage(imagePath);
    Texture2D::setDefaultAlphaPixelFormat(def);
    return texture;
}

NS_FLASHX_END /* NS_FLASHX_BEGIN */

File: swf-animate/api/FxAtlasManager.lua

--------------------------------
-- @module FxAtlasManager
-- @parent_module flashx

--------------------------------
-- 计算图集数量<br>
-- return 心跳单元数量
-- @function [parent=#FxAtlasManager] numAtlas 
-- @param self
-- @return unsigned short#unsigned short ret (return value: unsigned short)
        
--------------------------------
-- 删除未使用的图集
-- @function [parent=#FxAtlasManager] removeUnsuedAtlas 
-- @param self
-- @return FxAtlasManager#FxAtlasManager self (return value: flashx.FxAtlasManager)
        
--------------------------------
-- 删除所有缓存的图集
-- @function [parent=#FxAtlasManager] clear 
-- @param self
-- @return FxAtlasManager#FxAtlasManager self (return value: flashx.FxAtlasManager)
        
--------------------------------
-- 载入图集纹理<br>
-- param plistPath 图集配置路径<br>
-- param callback 回调
-- @function [parent=#FxAtlasManager] loadAtlasTexture 
-- @param self
-- @param #string plistPath
-- @param #function callback
-- @return FxAtlasManager#FxAtlasManager self (return value: flashx.FxAtlasManager)
        
--------------------------------
-- 释放一次图集<br>
-- param plistPath 图集配置路径
-- @function [parent=#FxAtlasManager] releaseAtlas 
-- @param self
-- @param #string plistPath
-- @return FxAtlasManager#FxAtlasManager self (return value: flashx.FxAtlasManager)
        
--------------------------------
-- 引用一次图集<br>
-- param plistPath 图集配置路径<br>
-- param texture 图集纹理
-- @function [parent=#FxAtlasManager] retainAtlas 
-- @param self
-- @param #string plistPath
-- @param #cc.Texture2D texture
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 使用设定的纹理格式载入图片<br>
-- param imagePath 图片路径<br>
-- param callback 回调<br>
-- param format 纹理格式<br>
-- return 纹理
-- @function [parent=#FxAtlasManager] loadImageByFormat 
-- @param self
-- @param #string imagePath
-- @param #function callback
-- @param #int format
-- @return Texture2D#Texture2D ret (return value: cc.Texture2D)
        
--------------------------------
-- 归还资源集<br>
-- param atlasArray 引用的动画资源表
-- @function [parent=#FxAtlasManager] releaseMuiltAtlas 
-- @param self
-- @param #array_table atlasArray
-- @return FxAtlasManager#FxAtlasManager self (return value: flashx.FxAtlasManager)
        
--------------------------------
-- 是否已载入图集<br>
-- param plistPath 图集配置路径<br>
-- return atlas
-- @function [parent=#FxAtlasManager] isAtlasLoaded 
-- @param self
-- @param #string plistPath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 
-- @function [parent=#FxAtlasManager] getInstance 
-- @param self
-- @return FxAtlasManager#FxAtlasManager ret (return value: flashx.FxAtlasManager)
        
return nil
